### DISCUSSION TITLE
Generating Grains and Lamellae

### DISCUSSION THREAD

**Discussion by janzenchoi** [neperfepx/neper#514] [Q&A] [answered]
Hello,

I have several questions regarding the generation of grains and lamellae, and was hoping that you guys would be able to answer them.

1) How do I specify the minimum and maximum of the grain equivalent diameter / sphericity distributions?
2) When defining the lamella widths externally, is it possible to define them based on the parent grain sizes? For example, is it possible to generate lamellae such that it equally divides the parent grain?
3) Can I define the grain equivalent diameters and sphericities externally (e.g., with fixed values, instead of via a distribution)?

Thank you very much for answering my questions, and thank you for providing and maintaining this incredible piece of software.

**Comment by rquey** [comment, marked as answer]
> Hello,
> 
> I have several questions regarding the generation of grains and lamellae, and was hoping that you guys would be able to answer them.
> 
>    How do I specify the minimum and maximum of the grain equivalent diameter / sphericity distributions?

See [-morpho](https://neper.info/doc/neper_t.html#cmdoption-morpho) which tells you this: 

> An interval of possible values can also be provided using interval(<min>,<max>).

So, you can use, for example, `neper -T -n 100 -morpho "diameq:interval(0.9,1.1),sphericity:interval(0.8,0.9)"`

>  When defining the lamella widths externally, is it possible to define them based on the parent grain sizes? For example, is it possible to generate lamellae such that it equally divides the parent grain?

Yes:
```
# Generate first scale and export grain diameqs
neper -T -n 100 -morpho gg  -statcell diameq

# Define widths based on diameqs, write as a mutliscale cell file
awk '{print NR,"lamellar(w="$1/10")"}' n100-id1.stcell > foo

# Generate 2-scale tessellation using previous tessellation for first scale and multiscale cell file for second scale
neper -T -n 100::from_morpho -morphooptiini "file(n100-id1.tess)" -morpho "voronoi::msfile(foo)" -o 2scale

# Visualize nicely
neper -V 2scale.tess -datacellcol "scaleid(1)" -datacelltrs 0.5 -print img
```
![image](https://user-images.githubusercontent.com/2659376/179469946-bd96b21b-39cb-40be-9bf4-242b0b2aa79c.png)

>   Can I define the grain equivalent diameters and sphericities externally (e.g., with fixed values, instead of via a distribution)?

See [-morpho](https://neper.info/doc/neper_t.html#cmdoption-morpho) which tells you this: 

> Cell-by-cell values can be provided using file(<file_name>), where <file_name> contains the cell values. 

```
# Generate random diameq values
gsl-randist 1 100 flat 0.8 1.2 > foo

# Generate tessellation
neper -T -n 100 -morpho "diameq:file(foo),1-sphericity:lognormal(0.145,0.03)" -o n100
```

> 
> Thank you very much for answering my questions, and thank you for providing and maintaining this incredible piece of software.

**Reply by janzenchoi** [reply]
Thank you very much @rquey. Your response was immensely helpful.

I do have another question, if you don't mind. I have several EBSD maps that I want to use to create an RVE. Currently, I am obtaining values for the grain sphericities and deriving the mean and variance values for the sphericity. Currently, the definition I am using for sphericity is shown below.

![image](https://user-images.githubusercontent.com/75835751/180033774-bec359c8-3079-49f1-a76a-7a6b59a5e33a.png)

Is this appropriate? If not, what definition for sphericity can I use with 2D EBSD maps. Additionally, what does the `1-` component of `1-sphericity` do, and how can I account for it when deriving the mean / variance values?

I am only asking because the grains that I am generating with my current mean / variance values are very abnormally shaped, resulting in terrible lamellae generation and long tessellation / mesh times.

**Reply by rquey** [reply]
I don't know if it is an appropriate expression, but it is not the one used in Neper. You either have to convert the values to match Neper's expression for *sphericity* or to implement this expression into Neper.

`1-sphericity` means that the distribution applies to 1 - sphericity; see [this paper](https://neper.info/papers.html#id2).

**Reply by janzenchoi** [reply]
Thanks @rquey.

I went through the paper, and I believe you guys defined sphericity as the "ratio between the surface area of the sphere of equivalent volume and the surface area of the grain". For a 2D map, do you have any ideas about how one could obtain the surface area of the grain? Would using circularity (i.e., ratio between the equivalent area and grain area) be sufficient?

Thanks again.

**Reply by mkasemer** [reply]
@janzennnnn The Neper documentation does not state what is being done in 2D. I believe, however, that your assumption needs a minor tweak: the 2D analog of sphericity would be the ratio between the circumference of the circle of equivalent area to the perimeter of the 2D grain. You can, of course, calculate this value yourself with a given 2D microstructure and compare to the values printed by Neper to confirm either assumption (mine or yours!).

**Reply by janzenchoi** [reply]
That makes a lot of sense. Thank you very much, @mkasemer.
